### PR TITLE
[DataGrid] Fix top corner pixels & outline radius

### PR DESCRIPTION
**Pull request by romgrk**
Closes: https://github.com/mui/mui-x/issues/11225

**Top pixels**:
image
image

**Outline radius (top-left & top-right) and bottom border**:
image

Before: https://codesandbox.io/p/sandbox/mui-mui-x-x-data-grid-forked-mmcydw
After: https://codesandbox.io/s/cocky-microservice-fcvj9h?file=/src/Demo.tsx

### CODE REVIEW

**Comment by Janpot** [comment]
I think you're referring to https://github.com/mui/mui-x/issues/11225. I think it's related, but doesn't fully cover it.

**Comment by romgrk** [comment]
I'll handle the outline issue separately.

File: packages/x-data-grid/src/components/GridPagination.tsx
Comment (cherniavskii) [inline packages/x-data-grid/src/components/GridPagination.tsx:1]: In the pagination, unwanted scrollbars have appeared: https://app.argos-ci.com/mui/mui-x/builds/22356/100923764

**Comment by romgrk** [comment]
Nevermind my comment above, let's deal with the outline here, it makes it easier to review both styling fixes if they're in one PR.

**Comment by romgrk** [comment]
Could someone with a macOS test & possibly tune this PR for that platform? The `border-top-right-radius` needs to be adjusted for the cases when there is a toolbar and when there is a vertical scrollbar, not sure if how I tuned it is correct if there are floating 0-width scrollbars like macOS has.

[Kooha-2024-07-25-17-54-42.webm](https://github.com/user-attachments/assets/02079c32-43e2-490a-9744-bf9f6868328b)

File: packages/x-data-grid/src/components/GridPagination.tsx
Comment (romgrk) [reply packages/x-data-grid/src/components/GridPagination.tsx:1]: I have fixed this for the cases where there shouldn't be any scrollbar.

However for the cases where there is an horizontal scrollbar, there will now be a vertical scrollbar on the pagination element. This is required if we want the `overflow: hidden` on the root container (which we need to avoid straight corners poking out of the border radius). Previously the pagination scrollbar would poke out of the pagination bounds, now it's going to be included in the pagination bounds.

File: packages/x-data-grid/src/components/GridRow.tsx [hunk 1]
```diff
@@ -62,40 +59,14 @@ export interface GridRowProps extends React.HTMLAttributes<HTMLDivElement> {
...
-type OwnerState = Pick<GridRowProps, 'selected'> & {
-  editable: boolean;
-  editing: boolean;
-  isFirstVisible: boolean;
-  isLastVisible: boolean;
-  classes?: DataGridProcessedProps['classes'];
-  rowHeight: GridRowProps['rowHeight'];
-};
-
-const useUtilityClasses = (ownerState: OwnerState) => {
```
Comment (romgrk) [inline packages/x-data-grid/src/components/GridRow.tsx:81]: I did a drive-by refactor here, not sure if it will be controversial. I was annoyed that I needed to add more stuff to the `type OwnerState` if I wanted to use variables internal to `GridRow`, so I removed this whole boilerplate section and replaced it with the `composeGridClasses` thing below that allows specifying utility classes inline. I could revert it if it's too controversial.

File: packages/x-data-grid/src/components/containers/GridRootStyles.ts
Comment (romgrk) [inline packages/x-data-grid/src/components/containers/GridRootStyles.ts:1]: The argos changes like [this one](https://app.argos-ci.com/mui/mui-x/builds/22361/100942132) are expected. The bottom border is now a real `border-bottom` instead of a floating `::after`, so it eats one pixel into the height of those header cells.

File: packages/x-data-grid/src/components/containers/GridRootStyles.ts [hunk 2]
```diff
@@ -273,6 +280,13 @@ export const GridRootStyles = styled('div', {
     [`& .${c.columnHeader}:focus, & .${c.cell}:focus`]: {
       outline: `solid ${t.palette.primary.main} 1px`,
     },
+    [`&.${c['root--noToolbar']} [aria-rowindex="1"] [aria-colindex="1"]`]: {
+      borderTopLeftRadius: 'var(--unstable_DataGrid-radius)',
+    },
+    [`&.${c['root--noToolbar']} [aria-rowindex="1"] [aria-colindex="${visibleColumns.length}"]`]: {
+      borderTopRightRadius:
+        'calc((1 - var(--DataGrid-hasScrollY)) * var(--unstable_DataGrid-radius))', // if (!hasScrollY) { style.borderTopRightRadius = radius }
```
Comment (cherniavskii) [inline packages/x-data-grid/src/components/containers/GridRootStyles.ts:288]: With the MacOS "When scrolling" scrollbar setting, the issue is only with the top-left demo:

https://github.com/user-attachments/assets/a025eb8e-f22b-48db-b48b-b32969ab32ec

I thought about utilizing the `--DataGrid-scrollbarSize` variable here, but couldn't find a way to do so.
But if we add a new 0/1 variable that would be set to 1 only if the scrollbar size is non-zero, that might work.
Thoughts?

File: packages/x-data-grid/src/components/containers/GridRootStyles.ts
Comment (KenanYusuf) [reply packages/x-data-grid/src/components/containers/GridRootStyles.ts:1]: There is an issue with this at the moment where the border does not span the whole grid when the header cells do not fill it. Is this a case where we need filler cells, or should we apply the border-bottom to the header row instead of the individual cells?

Screenshot 2024-07-26 at 11 52 04

File: packages/x-data-grid/src/components/containers/GridRootStyles.ts [hunk 3]
```diff
@@ -273,6 +280,13 @@ export const GridRootStyles = styled('div', {
...
+    },
```
Comment (KenanYusuf) [inline packages/x-data-grid/src/components/containers/GridRootStyles.ts:289]:
```suggestion
    [`&:not(:has(.${c.toolbarContainer})) [aria-rowindex="1"] [aria-colindex="1"]`]: {
      borderTopLeftRadius: 'var(--unstable_DataGrid-radius)',
    },
    [`&:not(:has(.${c.toolbarContainer}))[aria-rowindex="1"] [aria-colindex="${visibleColumns.length}"]`]:
      {
        borderTopRightRadius:
          'calc((1 - var(--DataGrid-hasScrollY)) * var(--unstable_DataGrid-radius))', // if (!hasScrollY) { style.borderTopRightRadius = radius }
      },
```

Could use a combination of `:not(:has())` here if it's preferable to adding a new class.

Comment (romgrk) [reply packages/x-data-grid/src/components/containers/GridRootStyles.ts:289]: I try to avoid non-class selectors when possible, more details in [these links](https://mui-org.slack.com/archives/C03SUL8TC1M/p1707306827856999). It's a balance to have with performance vs readability, but one important thing to note is that adding complex CSS selectors makes styling slower, because unlike JS which is run one line at a time, CSS selectors are always matched on all elements. Class selectors being the most common ones also means browsers optimize those heavily, but not so much other complex selectors.

In this particular case I'm not sure it would work though, the toolbar container is not in the hierarchy of the header cells IIRC.

File: packages/x-data-grid/src/components/containers/GridRootStyles.ts [hunk 4]
```diff
@@ -273,6 +280,13 @@ export const GridRootStyles = styled('div', {
     [`& .${c.columnHeader}:focus, & .${c.cell}:focus`]: {
       outline: `solid ${t.palette.primary.main} 1px`,
     },
+    [`&.${c['root--noToolbar']} [aria-rowindex="1"] [aria-colindex="1"]`]: {
+      borderTopLeftRadius: 'var(--unstable_DataGrid-radius)',
+    },
+    [`&.${c['root--noToolbar']} [aria-rowindex="1"] [aria-colindex="${visibleColumns.length}"]`]: {
```
Comment (cherniavskii) [inline packages/x-data-grid/src/components/containers/GridRootStyles.ts:286]: > I try to avoid non-class selectors when possible

How about using `columnHeader--last` instead of `[aria-colindex]`?

```suggestion
    [`&.${c['root--noToolbar']} [aria-rowindex="1"] .${c['columnHeader--last']}`]: {
```

File: packages/x-data-grid/src/components/containers/GridRootStyles.ts (hunk 3 above)
Comment (KenanYusuf) [reply packages/x-data-grid/src/components/containers/GridRootStyles.ts:289]: I tried it and it does work, but fair point on the performance and readability, it's probably not worth changing.

File: packages/x-data-grid/src/components/containers/GridRootStyles.ts (hunk 2 above)
Comment (romgrk) [reply packages/x-data-grid/src/components/containers/GridRootStyles.ts]: I handled it in javascript instead, could you test again and let me know if all looks good on macOS?

File: packages/x-data-grid/src/components/containers/GridRootStyles.ts (hunk 4 above)
Comment (romgrk) [reply packages/x-data-grid/src/components/containers/GridRootStyles.ts:288]: I thought about adding a class of that type to avoid JS logic in GridRootStyles but because we don't have a common wrapper for generic cells (including header cells and group header cells) I would have needed to replicate the logic in a few places, I opted for this instead.

Comment (romgrk) [reply packages/x-data-grid/src/components/containers/GridRootStyles.ts:288]: Wait no this won't work for group headers, their `colcount` is likely smaller than the headers `colcount`. Need to rework this.

File: packages/x-data-grid/src/components/containers/GridRootStyles.ts [hunk 5]
```diff
@@ -273,6 +282,15 @@ export const GridRootStyles = styled('div', {
...
+      borderTopRightRadius:
+        !dimensions.hasScrollY && dimensions.scrollbarSize > 0
+          ? 'var(--unstable_DataGrid-radius)'
+          : undefined,
```
Comment (cherniavskii) [inline packages/x-data-grid/src/components/containers/GridRootStyles.ts:292]:
```suggestion
        !dimensions.hasScrollY || dimensions.scrollbarSize === 0
          ? 'var(--unstable_DataGrid-radius)'
          : undefined,
```

I believe this is the correct condition.
With the suggestion above applied, seems to work fine:


https://github.com/user-attachments/assets/a4c7dbcb-7c44-4daa-b724-6bb7392f1a14

File: packages/x-data-grid/src/components/containers/GridRootStyles.ts (hunk 4 above)
Comment (romgrk) [reply packages/x-data-grid/src/components/containers/GridRootStyles.ts]: It's updated, I've added `columnHeader--last` on the groups.

File: packages/x-data-grid-pro/src/tests/rows.DataGridPro.test.tsx
Comment (romgrk) [inline packages/x-data-grid-pro/src/tests/rows.DataGridPro.test.tsx:1]: I've reviewed & approved argos changes, everything looks as expected. Interestingly, argos can pick up the 1px height offset for the text, but it can't pick up the missing pixels from the top corners that this PR fixes:

image

Comment (Janpot) [reply packages/x-data-grid-pro/src/tests/rows.DataGridPro.test.tsx:1]: They have a [threshold](https://argos-ci.com/docs/cypress#cyargosscreenshotname-options) setting for the sensitivity of their diff algorithm. We don't set it, but it's default at 0.5

File: packages/x-data-grid/src/components/containers/GridRootStyles.ts [hunk 6]
```diff
@@ -273,6 +280,15 @@ export const GridRootStyles = styled('div', {
     [`& .${c.columnHeader}:focus, & .${c.cell}:focus`]: {
       outline: `solid ${t.palette.primary.main} 1px`,
     },
+    [`&.${c['root--noToolbar']} [aria-rowindex="1"] [aria-colindex="1"]`]: {
+      borderTopLeftRadius: 'var(--unstable_DataGrid-radius)',
```
Comment (KenanYusuf) [inline packages/x-data-grid/src/components/containers/GridRootStyles.ts:284]:
```suggestion
      borderTopLeftRadius: 'calc(var(--unstable_DataGrid-radius) - 1px)',
```

There is a minor visual issue with using the same border radius for the element inside, in the screenshot below, you can see a gap between the curves:

Screenshot 2024-07-30 at 09 14 51

_I think_ we just want to minus a pixel based on this https://cloudfour.com/thinks/the-math-behind-nesting-rounded-corners/

**Review by KenanYusuf — APPROVED** [review]
Awesome work.

File: packages/x-data-grid/src/components/GridRow.tsx (hunk 1 above)
Comment (cherniavskii) [reply packages/x-data-grid/src/components/GridRow.tsx:81]: Why do you think it's controversial? I like it, can't see any downsides here 👍🏻

Comment (romgrk) [reply packages/x-data-grid/src/components/GridRow.tsx:81]: There is a very consistent convention throughout the company's codebase to use a split `useUtilityClasses` to define classes. Happy to see it's accepted.

File: packages/x-data-grid/src/hooks/features/columnHeaders/useGridColumnHeaders.tsx [hunk 7]
```diff
@@ -298,6 +300,7 @@ export const useGridColumnHeaders = (props: UseGridColumnHeadersProps) => {
         role="row"
         aria-rowindex={headerGroupingMaxDepth + 1}
         ownerState={rootProps}
+        className={clsx(pinnedRows.top.length === 0 && gridClasses['row--borderBottom'])}
```
Comment (cherniavskii) [inline packages/x-data-grid/src/hooks/features/columnHeaders/useGridColumnHeaders.tsx:303]: Nitpick: `clsx` feels like overkill here

**Review by cherniavskii — APPROVED** [review]
Looks great!

**Comment by cherniavskii** [comment]
I just noticed that the badge is being cut now:



I'm surprised this change was not detected by Argos diff: https://app.argos-ci.com/mui/mui-x/builds/22552/101656613

cc @romgrk
